01-intro/homework-01/hw-01.hs

```diff
@@ -0,0 +1,164 @@
+{-
+Validating Credit Card Numbers1
+Have you ever wondered how websites validate your credit card
+number when you shop online? They don’t check a massive database
+of numbers, and they don’t use magic. In fact, most credit providers
+rely on a checksum formula for distinguishing valid numbers from
+random collections of digits (or typing mistakes).
+In this section, you will implement the validation algorithm for
+credit cards. It follows these steps:
+
+• Double the value of every second digit beginning from the right.
+That is, the last digit is unchanged; the second-to-last digit is doubled; the third-to-last digit is unchanged; and so on. For example,
+[1,3,8,6] becomes [2,3,16,6].
+
+• Add the digits of the doubled values and the undoubled digits from the original number. For example, [2,3,16,6] becomes
+2+3+1+6+6 = 18.
+
+• Calculate the remainder when the sum is divided by 10. For the
+above example, the remainder would be 8.
+If the result equals 0, then the number is valid.
+-}
+
+
+{-
+Exercise 1 We need to first find the digits of a number. Define the
+functions
+toDigits :: Integer -> [Integer]
+toDigitsRev :: Integer -> [Integer]
+toDigits should convert positive Integers to a list of digits. (For 0 or
+negative inputs, toDigits should return the empty list.) toDigitsRev
+should do the same, but with the digits reversed.
+Example: toDigits 1234 == [1,2,3,4]
+Example: toDigitsRev 1234 == [4,3,2,1]
+Example: toDigits 0 == []
+Example: toDigits (-17) == []
+-}
+-- Exercise 1
+toDigits :: Integer -> [Integer]
+toDigits n = reverse (toDigitsRev n)
+
+toDigitsRev :: Integer -> [Integer]
+toDigitsRev n
+  | n <= 0 = []
+  | otherwise = (n `mod` 10) : toDigitsRev (n `div` 10)
+
+{-
+Exercise 2 Once we have the digits in the proper order, we need to
+double every other one. Define a function
+doubleEveryOther :: [Integer] -> [Integer]
+Remember that doubleEveryOther should double every other number beginning from the right, that is, the second-to-last, fourth-to-last,
+. . . numbers are doubled.
+Example: doubleEveryOther [8,7,6,5] == [16,7,12,5]
+Example: doubleEveryOther [1,2,3] == [1,4,3]
+-}
+-- Exercise 2
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther xs
+  | length (xs) `mod` 2 == 0 = doubleEven xs
+  | otherwise                = doubleOdd xs
+
+doubleEven :: [Integer] -> [Integer]
+doubleEven [] = []
+doubleEven (x:[]) = [2*x]
+doubleEven (x:y:zs) = 2*x : y : doubleEven zs
+
+doubleOdd :: [Integer] -> [Integer]
+doubleOdd [] = []
+doubleOdd (x:[]) = [x]
+doubleOdd (x:y:zs) = x : 2*y : doubleOdd zs
+
+{-
+Exercise 3 The output of doubleEveryOther has a mix of one-digit
+and two-digit numbers. Define the function
+sumDigits :: [Integer] -> Integer
+to calculate the sum of all digits.
+Example: sumDigits [16,7,12,5] = 1 + 6 + 7 + 1 + 2 + 5 = 22
+-}
+-- Exercise 3
+sumDigit :: Integer -> Integer
+sumDigit n
+  | n == 0    = 0
+  | otherwise = n `mod` 10 + sumDigit (n `div` 10)
```

Comment:
Guard clause does not seem necessary here.

```suggestion
-- Exercise 3
sumDigit :: Integer -> Integer
sumDigit 0 = 0
sumDigit n = n `mod` 10 + sumDigit (n `div` 10)
```

Reply:
@kwannoel Thanks!

Reply:
@kwannoel is there a preference to when not to use guard clauses?

Reply:
Using pattern matching here makes the code more succinct, and therefore more readable. 

If we use guard clause in this case, we add extra abstraction which we do not need. This reduces readability.

Guard clauses are generally used to express constraints:
``` haskell
f n | n < 0 = 0
    | otherwise = 1
```

In the above example, guard clauses / other branching abstractions are necessary because we can't express `n < 0` using [pattern matching](https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-580003.17).

```haskell
f (n < 0) = 0 -- ^ Parse error, we can only pattern match on values / variables
              -- (n < 0) only evaluates to a value 
f _ = 1
```

Reply:
@kwannoel That was a clear and efficient explanation, many thanks 💯